Connect: Add SearchBar

gzdunek opened the pull request:
**Remember to squash with co-authors before merging**

This PR adds an MVP of the search bar to Connect. Currently it's behind a feature flag (`feature.searchBar`) but we'll enable it by default before the release. The plan was to merge the code ASAP, potentially even straight to v12 when we have the chance, which didn't exactly pan out but there's no harm to having this feature flag for now.

The search works by essentially making a `ListResources` request for each supported resource type to every cluster the user is logged in to. We repurposed the old command palette UI for that but rewritten it to use React context and hooks rather than a class and a store. This allowed us to be a little bit more flexible as the old approach required every picker to conform to the same interface, both in terms of UI and code.

This implementation has two main pickers so far:

* `ActionPicker` which is the main one. It searches for resources but at the moment it also supports applying filters. In the future, we plan to add more actions to it such as "Open a new tab" or "Install tsh".
* `ParameterPicker` is activated when you pick an action from the `ActionPicker` that requires an additional parameter. Think choosing an SSH server or a db – you need to provide an SSH login or a db user for those item. In those situations, `ActionPicker` will switch to `ParameterPicker` and let you pick a relevant item from the list. 

Everything is contained within `web/packages/teleterm/src/ui/Search`. Arguably, `useSearch` could be refactored a little bit to maybe make its structure a little more clear as it handles both the resource search and the filter search. However, at the moment we're not totally sure how the search bar will evolve, so we want to leave any bigger refactors for later. We added a couple of basic tests for regressions that happened so far. We also have stories for the items from the action picker.

The next steps are outlined in [the search bar doc](https://docs.google.com/document/d/1wMurmpEIHg2Wiugs1AQAqnBUU_YkIfNj6e-lVaHKKuI/edit#heading=h.6gzk7romrhe9). Error handling will be added in an upcoming PR. Docs updates will be done in a separate PR as well.

zmb3 commented on web/packages/shared/hooks/useAsync.ts:
```diff
@@ -188,3 +188,23 @@ export function makeErrorAttempt<T>(statusText: string): Attempt<T> {
+/**
+ * mapAttempt maps attempt data but only if the attempt is successful.
+ */
+export function mapAttempt<A, B>(
```
```suggestion
export function mapSuccessfulAttempt<A, B>(
```

zmb3 commented on web/packages/teleterm/src/services/config/appConfigSchema.ts:
```diff
@@ -68,6 +72,9 @@ export const createAppConfigSchema = (platform: Platform) => {
     'keymap.newTab': shortcutSchema
       .default(defaultKeymap['newTab'])
       .describe(getShortcutDesc('open a new tab')),
+    'keymap.newTerminalTab': shortcutSchema
```
Is this related to the search bar in any way? If not, I would move it to a separate PR.

ravicious replied:
It is related! Before the search bar was introduced, you could open a new terminal tab by pressing Cmd + K, typing any command you want and then pressing Enter.

With the addition of the search bar, there's no way at the moment to open a local terminal tab from the search bar. So we added a new item to the More Options menu to the left of the profile selector. The shortcut is displayed there to the right of the action. This way the user can still open a new terminal tab using the keyboard only.

Since the plan is to backport the search bar to v12.3, I say we can just keep those commits in this PR.

ravicious committed:
Connect: Add SearchBar

This commit adds an MVP of the search bar to Connect. Currently it's
behind a feature flag (`feature.searchBar`) but we'll enable it by
default before the release. The plan was to merge the code ASAP,
potentially even straight to v12 when we have the chance, which didn't
exactly pan out but there's no harm to having this feature flag for now.

On top of that, this commit adds a new shortcut to open the search bar
(this replaces the current shortcut to open the command bar) and a
shortcut to open a new terminal tab.

The search works by essentially making a `ListResources` request for
each supported resource type to every cluster the user is logged in to.
We repurposed the old command palette UI for that but rewritten it to
use React context and hooks rather than a class and a store. This
allowed us to be a little bit more flexible as the old approach required
every picker to conform to the same interface, both in terms of UI and
code.

This implementation has two main pickers so far:

* `ActionPicker` which is the main one. It searches for resources but at
  the moment it also supports applying filters. In the future, we plan
  to add more actions to it such as "Open a new tab" or "Install tsh".
* `ParameterPicker` is activated when you pick an action from the
  `ActionPicker` that requires an additional parameter. Think choosing
  an SSH server or a db – you need to provide an SSH login or a db user
  for those item. In those situations, `ActionPicker` will switch to
  `ParameterPicker` and let you pick a relevant item from the list.

Everything is contained within `web/packages/teleterm/src/ui/Search`.
Arguably, `useSearch` could be refactored a little bit to maybe make its
structure a little more clear as it handles both the resource search and
the filter search. However, at the moment we're not totally sure how the
search bar will evolve, so we want to leave any bigger refactors for
later. We added a couple of basic tests for regressions that happened so
far. We also have stories for the items from the action picker.

Error handling will be added in an upcoming PR. Docs updates will be
done in a separate PR as well.

Co-authored-by: Grzegorz Zdunek <grzegorz.zdunek@goteleport.com>
